index.js

```diff
@@ -147,43 +167,47 @@ module.exports = async function (url, options) {
-
-                if (!retry(retryOptions, null, response)) {
-                    // response.timeout should reflect the actual timeout
-                    response.timeout = retryOptions.socketTimeout;
-                    return resolve(response);
-                }
-
-                console.error(`Retrying in ${retryOptions.retryInitialDelay} milliseconds, attempt ${attempt - 1} failed (status ${response.status}): ${response.statusText}`);
-            } catch (error) {
-                clearTimeout(timeoutHandler);
-
-                if (!retry(retryOptions, error, null)) {
-                    if (error.name === 'AbortError') {
-                        return reject(new FetchError(`network timeout at ${url}`, 'request-timeout'));
+            while (!isResponseTimedOut(retryOptions)) {
+                ++attempt;
+                const waitTime = getRetryDelay(retryOptions);
+
+                let timeoutHandler;
+                if (retryOptions.socketTimeout) {
+                    const controller = new AbortController();
+                    timeoutHandler = setTimeout(() => controller.abort(), retryOptions.socketTimeout);
+                    options.signal = controller.signal;
+                }                
+    
+                try {
+                    const response = await fetch(url, options);
+
+                    if (shouldRetry(retryOptions, null, response, waitTime)) {
+                        console.error(`Retrying in ${waitTime} milliseconds, attempt ${attempt} failed (status ${response.status}): ${response.statusText}`);
+                    } else {
+                        // response.timeout should reflect the actual timeout
+                        response.timeout = retryOptions.socketTimeout;
+                        return resolve(response);
                     }
-
-                    return reject(error);
+                } catch (error) {
+                    const response = {status: error.code || 500, statusMessage : error.message || "Unknown server error"};
+                    if (!shouldRetry(retryOptions, error, response, waitTime)) {
+                        if (error.name === 'AbortError') {
+                            return reject(new FetchError(`network timeout at ${url}`, 'request-timeout'));
+                        } else {
+                            return reject(error);
+                        }
+                    }
+                    console.error(`Retrying in ${waitTime} milliseconds, attempt ${attempt} error: ${error.message}`);
+                } finally {
+                    clearTimeout(timeoutHandler);
                 }
-
-                console.error(`Retrying in ${retryOptions.retryInitialDelay} milliseconds, attempt ${attempt - 1} error: ${error.message}`);
+                // Fetch loop is about to repeat, delay as needed first.
+                if (waitTime > 0) {
+                    await new Promise(resolve => setTimeout(resolve, waitTime));
```

Comment:
I am curious what happens if this timeout resolves during the retry request

Reply:
Right now there's no test asserting an obvious timeout issue but if needed we could wrap with a try/catch and reject with the caught error.  I'd rather reproduce the issue first rather than introduce vestigial code though.

Reply:
+1 ^^^^^